amd/src/editor.js

```diff
@@ -0,0 +1,993 @@
+        } else {
+            // This case is user select a text that is not inside subsup.
+            //  We retrieve the selected text and then delete it in DOM.
+            const selectedText = range.toString();
+            range.deleteContents();
+            const previousNode = range.commonAncestorContainer.previousSibling;
+            const nextNode = range.commonAncestorContainer.nextSibling;
+            // In addition, we will merge adjacent sup/sub tags into a single sup/sub tag.
+            if (previousNode || nextNode) {
+                const newNode = this.createElement(tag);
+                let startOffset = 0;
+                let endOffset = 0;
+                let content = '';
+                if (previousNode && previousNode?.nodeName?.toLowerCase() === tag) {
+                    content = previousNode.textContent;
+                    startOffset = content.length;
+                    previousNode.remove();
+                }
+                content += selectedText;
+                endOffset = content.length;
+                if (nextNode && nextNode?.nodeName?.toLowerCase() === tag) {
+                    content += nextNode.textContent;
+                    nextNode.remove();
+                }
+                newNode.textContent = content;
+                if (content !== selectedText) {
+                    range.insertNode(newNode);
+                    range.setStart(newNode.firstChild, startOffset);
+                    range.setEnd(newNode.firstChild, endOffset);
+                    this.getTextArea().value = this.getCleanHTML();
+                    return;
+                }
+            }
+
+            // Create a sup/sub tag that wrap the selected text.
+            const newNode = document.createElement(tag);
+            newNode.appendChild(document.createTextNode(selectedText));
+            // Make a selection to the selected text.
+            selection.removeAllRanges();
+            // Insert it into DOM.
+            range.insertNode(newNode);
+            range.selectNodeContents(newNode.firstChild);
+            selection.addRange(range);
+            // Clean up all the empty text.
+            this.getEditorContent().childNodes.forEach(el => {
+                if (el.nodeName === '#text' && el.textContent === '') {
+                    el.remove();
+                }
+            });
+            this.getTextArea().value = this.getCleanHTML();
+        }
+        // Clean up.
+        this.getEditorContent().childNodes.forEach(el => {
+            if (el.nodeName === '#text' && el.textContent === '') {
+                el.remove();
+            }
+        });
+        this.saveHistory();
+    }
+
+    /**
+     * Save history for undo/redo actions.
+     */
+    saveHistory() {
+        const content = this.getCleanHTML();
+        if (this.historyIndex === -1 || content !== this.history[this.historyIndex]) {
+            this.history.splice(this.historyIndex + 1);
```

Comment:
Not sure if we need the below line of code?
 this.history.splice(this.historyIndex + 1);

Reply:
Sorry @AnupamaSarjoshi :(,
I affair that I can't remove that line of code due to:

The line this.history.splice(this.historyIndex + 1); is important because it makes sure that when a new edit is made after using "Undo," the "Redo" options are cleared. Let’s break it down with an example:

**Without splice:**
Initial Edits:
The user types: edit1, edit2, edit3.
History: [edit1, edit2, edit3]
Index: 2 (pointing to edit3).

Undo Action:
The user undoes once.
History: [edit1, edit2, edit3]
Index: 1 (pointing to edit2).

New Edit:
The user types edit4.
History (without splice): [edit1, edit2, edit3, edit4]
This is incorrect because edit3 is no longer relevant but still exists in the history. The undo/redo behavior will become confusing.

With splice:
Initial Edits:
The user types: edit1, edit2, edit3.
History: [edit1, edit2, edit3]
Index: 2.

Undo Action:
The user undoes once.
History: [edit1, edit2, edit3]
Index: 1.

New Edit:
The user types edit4.
The line this.history.splice(this.historyIndex + 1); clears everything after edit2.
History (with splice): [edit1, edit2, edit4]
Now the history is correct and matches what the user expects.

Why Keep It?
Without splice, the history can keep old edits (edit3), which makes undo/redo confusing. By keeping splice, the history stays linear and predictable, like most users expect.

Thanks.

Reply:
Thanks @BruceGoodGuy, for explaining in detail. That's helpful :)